lib/utils.js

```diff
@@ -31,3 +36,105 @@ function throwPromiseNotDefined() {
     'Your Node runtime does support ES6 Promises. ' +
     'Set "global.Promise" to your preferred implementation of promises.');
 }
+
+/**
+ * Divide an async call with large array into multiple calls using smaller chunks
+ * @param {Array} largeArray - the large array to be chunked
+ * @param {Number} chunkSize - size of each chunks
+ * @param {Function} processFunction - the function to be called multiple times
+ * @param {Function} cb - the callback
+ */
+function uploadInChunks(largeArray, chunkSize, processFunction, cb) {
+  var self = this;
+  var chunkArrays = [];
+
+  if (largeArray.length > chunkSize) {
+    // copying so that the largeArray object does not get affected during splice
+    var copyOfLargeArray = [].concat(largeArray);
+
+    // chunking to smaller arrays
+    while (copyOfLargeArray.length > 0) {
+      chunkArrays.push(copyOfLargeArray.splice(0, chunkSize));
+    }
+
+    var tasks = chunkArrays.map(function(chunkArray) {
+      return function(previousResults, chunkCallback) {
+        var lastArg = arguments[arguments.length - 1];
+
+        if (typeof lastArg === 'function') {
+          chunkCallback = lastArg;
+        }
+
+        processFunction.call(self, chunkArray, function(err, results) {
+          if (err) {
+            return chunkCallback(err);
+          }
+
+          // if this is the first async waterfall call or if previous results was not defined
+          if (typeof previousResults === 'function' || typeof previousResults === 'undefined' ||
+            previousResults === null) {
+            previousResults = results;
+          } else if (results) {
+            previousResults = concatResults(previousResults, results);
+          }
+
+          chunkCallback(err, previousResults);
+        });
+      };
+    });
+
+    async.waterfall(tasks, cb);
+  } else {
+    processFunction.call(self, largeArray, cb);
```

Comment:
Short "else" blocks after a long "if" block make it difficult to build a full picture. Please use reverse-logic with early-return instead:

```js
if (largeArray.length <= chunkSize) {
  return processFunction.call(self, largeArray, cb);
}

// the rest of the code is indented one level less
```

Also IIRC, `fn.call(...)` method has performance overhead compared to regular function calls `fn(...)`. Also AFAICT, you are not using `this` in any of the callbacks. Therefore I would prefer to use `processFunction(largeArray, cb)`. Unless I am missing something?